packages/marshal/src/encodeToSmallcaps.js

```diff
@@ -0,0 +1,452 @@
+   * above, i.e., `decodeFromSmallcaps` cannot rely on `encoding` being a
+   * valid marshalled representation. Rather, `decodeFromSmallcaps` must
+   * validate that.
+   *
+   * @param {SmallcapsEncoding} encoding must be hardened
+   */
+  const decodeFromSmallcaps = encoding => {
+    switch (typeof encoding) {
+      case 'boolean':
+      case 'number': {
+        return encoding;
+      }
+      case 'string': {
+        if (encoding === '') {
+          return encoding;
+        }
+        switch (encoding[0]) {
+          case "'": {
+            // un-hilbert-ify the string
+            return encoding.slice(1);
+          }
+          case '%': {
+            return passableSymbolForName(encoding.slice(1));
+          }
+          case '#': {
+            switch (encoding) {
+              case '#undefined': {
+                return undefined;
+              }
+              case '#NaN': {
+                return NaN;
+              }
+              case '#Infinity': {
+                return Infinity;
+              }
+              case '#-Infinity': {
+                return -Infinity;
+              }
+              default: {
+                assert.fail(X`unknown constant "${q(encoding)}"`, TypeError);
+              }
+            }
+          }
+          case '+':
+          case '-': {
+            return BigInt(encoding);
+          }
+          case '$': {
+            const result = decodeRemotableFromSmallcaps(encoding);
+            if (passStyleOf(result) !== 'remotable') {
+              assert.fail(
+                X`internal: decodeRemotableFromSmallcaps option must return a remotable: ${result}`,
+              );
+            }
+            return result;
+          }
+          case '&': {
+            const result = decodePromiseFromSmallcaps(encoding);
+            if (passStyleOf(result) !== 'promise') {
+              assert.fail(
+                X`internal: decodePromiseFromSmallcaps option must return a promise: ${result}`,
+              );
+            }
+            return result;
+          }
+          case '!':
+          case '"':
+          case '(':
+          case ')':
+          case '*':
+          case ',': {
+            assert.fail(
+              X`Special char ${q(
+                encoding[0],
+              )} reserved for future use: ${encoding}`,
+            );
+          }
+          default: {
+            return encoding;
+          }
+        }
+      }
+      case 'object': {
+        if (encoding === null) {
+          return encoding;
+        }
+
+        if (isArray(encoding)) {
+          const result = [];
+          const { length } = encoding;
+          for (let i = 0; i < length; i += 1) {
+            result[i] = decodeFromSmallcaps(encoding[i]);
+          }
+          return result;
+        }
+
+        if (hasOwnPropertyOf(encoding, '#tag')) {
+          const { '#tag': tag, payload, ...rest } = encoding;
```

Comment:
I haven't looked at the tests yet, but we will want to have one verifying that an object with both "#tag" and "#error" properties is rejected to ensure that a refactoring here does not break anything.

Reply:
Done